File: GuardeSoftwareAPI/Dtos/Communication/CommunicationDto.cs
namespace GuardeSoftwareAPI.Dtos.Communication;

public class CommunicationDto
{
    public int Id { get; set; }
    public string Title { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
    public string? SendDate { get; set; } // yyyy-MM-dd
    public string? SendTime { get; set; } // HH:mm
    public string Channel { get; set; } = string.Empty; // "Email", "WhatsApp", "Email + WhatsApp"
    public List<string> Recipients { get; set; } = [];
    public string Status { get; set; } = string.Empty; // "Sent", "Scheduled", "Draft"
    public string CreationDate { get; set; } = string.Empty; // yyyy-MM-dd
    public int? SmtpConfigId { get; set; }
    public bool IsAccountStatement { get; set; }
}
    
    public class UpsertCommunicationRequest
{
    public int? Id { get; set; }
    public string Title { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
    public string? SendDate { get; set; }
    public string? SendTime { get; set; }
    public string Type { get; set; } = string.Empty; // 'schedule' | 'draft'
    public int? SmtpConfigId { get; set; }
    public List<string> Channels { get; set; } = [];
    public List<string> Recipients { get; set; } = [];
    public List<IFormFile>? Attachments { get; set; }
    public bool IsAccountStatement { get; set; }
}

File: GuardeSoftwareClient/src/app/core/dtos/communications/communicationDto.ts
export interface CommunicationDispatchDto {
  dispatchId: number;
  clientId: number;
  externalRecipientId?: number | null;
  isExternalRecipient?: boolean;
  clientName: string;
  channel: string;
  status: string;
  errorMessage: string;
  dispatchDate: string;
  recipientPhone?: string | null;
  isTest?: boolean;
  isSelected?: boolean;
  hasContent?: boolean;
}

export type CommunicationExtensionMode = 'never-attempted' | 'without-success';

export interface CommunicationExtensionRecipient {
  id: number;
  name: string;
  email: string;
  type: string | null;
  isActive: boolean;
  isAssociated: boolean;
  hasRealAttempt: boolean;
  hasRealSuccess: boolean;
  lastStatus: string | null;
  lastAttemptWasTest: boolean;
}

export interface CommunicationExtensionPreview {
  communicationId: number;
  title: string;
  status: string;
  recipientType: string;
  mode: CommunicationExtensionMode;
  totalInDirectory: number;
  eligibleWithEmail: number;
  alreadySuccessfulCount: number;
  neverAttemptedCount: number;
  previouslyAttemptedCount: number;
  failedOrPendingCount: number;
  alreadyAssociatedCount: number;
  newToCommunicationCount: number;
  selectedForSendCount: number;
  inactiveOrWithoutEmailCount: number;
  isTestCommunication: boolean;
  candidateListTruncated: boolean;
  recipients: CommunicationExtensionRecipient[];
}

export interface ExtendCommunicationRequest {
  recipientType: string;
  mode: CommunicationExtensionMode;
}

export interface CommunicationExtensionResult extends CommunicationExtensionPreview {
  queued: boolean;
  addedAssociationCount: number;
  communication?: ComunicacionDto | null;
}

export interface CommunicationExternalRecipientDto {
  id: number;
  name: string | null;
  email: string | null;
  type: string | null;
}

export interface ComunicacionDto {
  id: number;
  title: string;
  content: string;
  sendDate: string | null;
  sendTime: string | null;
  channel: string;
  recipients: string[];
  status: 'Draft' | 'Scheduled' | 'Procesando' | 'Finished' | 'Finished w/ Errors' | 'Failed';
  creationDate: string;
  smtpConfigId?: number | null;
  isAccountStatement: boolean;
  isNextMonthStatement: boolean;
  sendToAllEmails: boolean;
  externalRecipients: CommunicationExternalRecipientDto[];
  errorMessage?: string | null;
  dispatches?: CommunicationDispatchDto[];
}


export interface UpsertComunicacionRequest {
  id: number | null;
  title: string;
  content: string;
  sendDate: string | null;
  sendTime: string | null;
  channels: ('Email' | 'WhatsApp')[];
  recipients: string[];
  type: 'schedule' | 'draft';
  smtpConfigId?: number | null;
  isAccountStatement: boolean;
  isNextMonthStatement: boolean;
  sendToAllEmails: boolean;
  externalRecipientIds: number[];
}
